### DISCUSSION TITLE
Make auto-pair config additive

### DISCUSSION THREAD

**Discussion by NoraPerry** [helix-editor/helix#4494] [Ideas]
I want to disable the ```'`'``` auto-pair in my config but as far as I can tell, the only way to do that is to copy all the default auto-pairs into my config and remove the line for ```'`'``` like this:

```
[editor.auto-pairs]
'(' = ')'
'{' = '}'
'[' = ']'
'"' = '"'
'<' = '>'
```

I don't like this because it's impossible to know what this config section is actually doing without a comment, and if any new pairs are added to the default config I won't get them unless I keep track of the changelog and update my config.

Instead, I think it would be better to have new pairs in the user's config be added to the default list rather than overwriting it, and there should be some syntax (e.g. ```'`' = false```) to remove a pair.

**Comment by dead10ck** [comment]
This sounds complicated to me personally. Is it that important that if the defaults are updated that you get them transparently? I can't imagine this will happen very often.

Also the flip side to this approach is that it's hard to piece together what all the pairs are, where with how it is now, what you see is what you get.

**Reply by NoraPerry** [reply]
> This sounds complicated to me personally. Is it that important that if the defaults are updated that you get them transparently? I can't imagine this will happen very often.

You're right I don't think the default pairs will change often, but I think additive options would be useful for other things not just this specific case.

If I have a list of custom pairs in `config.toml` then I want to add some more custom pairs for specific languages in `languages.toml`, I would have to copy the entire list from the global config and paste it in for every language then remember to keep everything in sync. That's just an example, not something I've run into in my config as I don't currently need a `languages.toml` but I don't think it would be super uncommon for people to want to do these kinds of things with their configs.

> Also the flip side to this approach is that it's hard to piece together what all the pairs are, where with how it is now, what you see is what you get.

This is a fair point but I think it's unlikely that someones config would be mutating the options so much that it's genuinely hard to figure out what's going on, especially as the config file is usually personalized by the specific user and not a file that random people will often need to read and understand.

Personally, when reading/writing my config I think it's harder to understand with the current approach because what I see is "a bunch of pairs are listed" and what I get is "one of the default pairs is removed", but this is totally a matter of perspective and it could be argued either way.

**Comment by the-mikedavis** [comment]
It could be nice to have some config like this changeable without specifying the orginal value in full. But I don't think we should approach it until we use a different config language than TOML. Something with scripting support that could take the previous value as an argument and explicitly add or subtract elements would be preferrable. Building something like that in a TOML DSL is not worth the complexity that it adds IMO.

**Reply by dead10ck** [reply]
Yeah my thoughts exactly. TOML makes this really complicated.
